File: okhttp-tests/src/test/java/okhttp3/XxX.java
package okhttp3;

import java.io.IOException;
import java.net.InetSocketAddress;
import java.net.Proxy;
import java.net.Socket;
import java.util.concurrent.TimeUnit;
import javax.net.SocketFactory;

public class XxX {
  public static void main(String[] args) throws IOException {
    SocketFactory socketFactory = new MyDelegatingSocketFactory();
    OkHttpClient.Builder builder = new OkHttpClient.Builder();
    builder.proxy(new Proxy(Proxy.Type.SOCKS, new InetSocketAddress("localhost", 2001)));
    builder.connectTimeout(10, TimeUnit.SECONDS);
    builder.socketFactory(socketFactory);
    OkHttpClient client = builder.build();

    Request req = new Request.Builder().url("https://google.com/robots.txt").build();
    try (Response response = client.newCall(req).execute()) {
      System.out.println(response.body.string().length());
    }
  }

  public static class MyDelegatingSocketFactory extends DelegatingSocketFactory {
    public MyDelegatingSocketFactory() {
      super(SocketFactory.getDefault());
    }

    @Override public Socket configureSocket(Socket socket) throws IOException {
      // set android
      //TrafficStats.tagSocket(socket);

      System.out.println("configured");

      return socket;
    }
  }
}

File: okhttp/src/main/kotlin/okhttp3/XxX.kt
package okhttp3

import java.io.IOException
import java.util.concurrent.CountDownLatch
import java.util.concurrent.TimeUnit

fun main() {
  val client = OkHttpClient.Builder()
    .pingInterval(2500, TimeUnit.MILLISECONDS)
    .connectionPool(ConnectionPool(0, 1, TimeUnit.NANOSECONDS))
    .build()

  val latch = CountDownLatch(1)

  val request = Request.Builder().url("https://nghttp2.org/httpbin/get").build()
  client.newCall(request).enqueue(object : Callback {
    override fun onFailure(call: Call, e: IOException) {
      println("onFailure")
      latch.countDown()
    }

    override fun onResponse(call: Call, response: Response) {
      println("onResponse")
//      println(response.body?.string())
      println(response.protocol)
//      response.close()
      latch.countDown()
    }
  })

  println("await")
  latch.await()
  println("awaited")

  println(client.connectionPool.connectionCount())
//  client.connectionPool.evictAll()
//  println(client.connectionPool.connectionCount())
}
